tests/test_wheel.py

```diff
@@ -45,3 +52,62 @@ def test_find_metadata_files():
     ]
     candidates = wheel.Wheel.find_candidate_metadata_files(names)
     assert expected == candidates
+
+
+def test_read_valid(example_wheel):
+    """Test reading a valid wheel file"""
+    metadata = example_wheel.read().decode().splitlines()
+    assert "Name: twine" in metadata
+    assert "Version: 1.5.0" in metadata
+
+
+def test_read_non_existent_wheel_file_name():
+    """Test reading a wheel file which doesn't exist"""
+
+    file_name = "/foo/bar/baz.whl"
+    with pytest.raises(exceptions.InvalidDistribution) as excinfo:
+        wheel.Wheel(file_name)
+
+    assert "No such file: {}".format(file_name) == str(excinfo.value)
+
+
+def test_read_invalid_wheel_extension():
+    """Test reading a wheel file without a .whl extension"""
+
+    file_name = os.path.join(os.path.dirname(__file__), "fixtures/twine-1.5.0.tar.gz")
+    with pytest.raises(exceptions.InvalidDistribution) as excinfo:
+        wheel.Wheel(file_name)
+
+    assert "Not a known archive format: {}".format(file_name) == str(excinfo.value)
+
+
+def test_read_wheel_without_metadata(capsys, tmpdir):
+    """Test reading a wheel file with empty metadata"""
+
+    def create_empty_metadata_wheel(name, version):
```

Comment:
Looks like there's a lot going on here, and to my eye neither `create_empty_metadata_wheel` nor `add_file` need to be nested. I'm curious why you chose to write them that way.

I haven't yet taken the time to understand what this code does, but I wonder if there's a simpler/clearer way to accomplish the desired outcome. For example, could we add a wheel without metadata to `tests/fixtures`?

Reply:
So I actually took the idea of creating an empty wheel from how [pip](https://github.com/pypa/pip/blob/master/tests/lib/__init__.py#L948) does it.

This actually creates a temporary wheel file, but with empty metadata as you correctly understood. I thought that this is better left to a function, since this is a one off test and might not need a special wheel without metadata, unless other tests end up using it, which I am not sure will be the case.

Reply:
Thanks for explaining. I'd still probably move it out into it's own function. Also, when copying (or borrowing heavily) code from another source, I think it's helpful for future readers to cite a URL in the docstring.

However! After looking at the source for [`wheel.Wheel.read`](https://github.com/pypa/twine/blob/f61707b93a4f5dd89dfdd22db1acb846b8932b9c/twine/wheel.py#L54-L78), I don't think this function is necessary. It looks like we can trigger the `METADATA` error by using any zip file that has ` .whl` extension, but not the right contents.

I checked out this branch locally, and created an empty wheel:

```
$ touch not-a-wheel && zip -m not-a-wheel.whl not-a-wheel
```

Then changed the test to:

```python
    whl_file = os.path.join(os.path.dirname(__file__), "fixtures/not-a-wheel.whl")

    with pytest.raises(
        exceptions.InvalidDistribution, match=f"No METADATA in archive: {whl_file}"
    ):
        wheel.Wheel(whl_file)
```

which passes.

How does that feel to you?

Reply:
Thanks for the detailed suggestion. I have created a empty whl file containing an empty `METADATA` file for this test using the `tmpdir` fixture